### PR TITLE
fix: refresh cached header when we set request header

### PR DESCRIPTION
**Pull request by spacewander**
### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/core/request.lua [hunk 1]
```diff
@@ -68,12 +68,25 @@ function _M.header(ctx, name)
 end
 
 
-function _M.set_header(header_name, header_value)
+function _M.set_header(ctx, header_name, header_value)
+    if type(ctx) == "string" then
+        -- It would be simpler to keep compatibility if we put 'ctx'
+        -- after 'header_value', but the style is too ugly!
```
Comment (membphis) [inline apisix/core/request.lua:74]: how about print a warning message to log?

**Comment by jenskeiner** [comment]
@spacewander I reckon this would obsolete a similar change I made here: bfa4aa7297c14c181a7403daea4afdce2684040f Correct?

I had noticed that setting a request header in `openid-connect` plugin would not be visible to the downstream `auth-keycloak` plugin since the headers there are retrieved from a cache which hadn't been updated.

Better to fix in the core method anyway, so +1 for this change.

**Comment by spacewander** [comment]
@jenskeiner 
Yes, you are right!

File: apisix/core/request.lua (hunk 1 above)
Comment (spacewander) [reply apisix/core/request.lua:74]: @membphis 
Added.

**Review by moonming — APPROVED** [review]

**Review by membphis — APPROVED** [review]

**Pull request merged**